backend/proxy_server/github.go

```diff
@@ -0,0 +1,166 @@
+	Blog              string `json:"blog"`
+	Location          string `json:"location"`
+	Email             string `json:"email"`
+	Hireable          bool   `json:"hireable"`
+	Bio               string `json:"bio"`
+	TwitterUserName   string `json:"twitter_username"`
+	PublicRepos       int    `json:"public_repos"`
+	PublicGists       int    `json:"public_gits"`
+	Followers         int    `json:"followers"`
+	Following         int    `json:"following"`
+	CreatedAt         string `json:"created_at"`
+	UpdatedAt         string `json:"updated_at"`
+}
+
+// deal with github user login
+func githubLoginHandler(c *gin.Context) {
+
+	postCode := &struct{ Code string }{
+		Code: "",
+	}
+	if err := c.BindJSON(postCode); err != nil {
+		JSONErr(c, http.StatusBadRequest, fmt.Sprintf("login: get authorization code fail: %v", err))
+		return
+	}
+
+	token, err := getOauthToken(githubConfig, postCode.Code)
+	if token == nil || token.AccessToken == "" {
+		logger.Warn(c.ClientIP(), "get token fail", err.Error())
+		JSONErr(c, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	githubUserInfo, err := getUserGithubInfo(token)
+	if githubUserInfo == nil {
+		logger.Warn(c.ClientIP(), "get github user info fail", err.Error())
+		JSONErr(c, http.StatusInternalServerError, "get github user info fail")
+		return
+	}
+
+	localUser, err := getLocalUser(githubUserInfo)
+	if localUser == nil {
+		logger.Warn(c.ClientIP(), "get local user info fail", err.Error())
+		JSONErr(c, http.StatusInternalServerError, "get local user info fail")
+		return
+	}
+	c.JSON(http.StatusOK, localUser)
+}
+
+// exchange authorization code to token
+func getOauthToken(githubconfig *oauth2.Config, code string) (*oauth2.Token, error) {
+
+	var token oauth2.Token
+
+	reqURL := "https://github.com/login/oauth/access_token?" + "client_id=" +
+		githubconfig.ClientID + "&client_secret=" + githubconfig.ClientSecret +
+		"&code=" + code
+	req, err := http.NewRequest(http.MethodPost, reqURL, nil)
+	if err != nil {
+		return nil, errors.New("internal error, please try again")
+	}
+	req.Header.Set("accept", "application/json")
+
+	httpClient := http.Client{Timeout: 5 * time.Second}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	err = json.NewDecoder(resp.Body).Decode(&token)
+	if err != nil {
+		return nil, err
+	}
+	return &token, errors.New("exchange token success")
+
+}
+
+// get userinfo from api.github.com/user
+func getUserGithubInfo(token *oauth2.Token) (*githubUser, error) {
+
+	httpClient := githubConfig.Client(context.TODO(), token)
+	userInfo, err := httpClient.Get("https://api.github.com/user")
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer userInfo.Body.Close()
+
+	info, err := ioutil.ReadAll(userInfo.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var githubUserInfo githubUser
+	err = json.Unmarshal(info, &githubUserInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &githubUserInfo, errors.New("get github user info success")
```

Comment:
same as above

Reply:
My falut. I should have checked all the code.